app/questionnaire/questionnaire_schema.py

```diff
@@ -848,17 +870,60 @@ def _populate_when_rules_section_dependencies(self) -> None:
             when_rules = self.get_values_for_key(section, "when")
             rules: list = list(when_rules)
 
-            if rules_section_dependencies := self._get_rules_section_dependencies(
-                section["id"], rules
-            ):
+            (
+                rules_section_dependencies,
+                rule_section_dependencies_for_progress_value_source,
+            ) = self._get_rules_section_dependencies(section["id"], rules)
+
+            if rules_section_dependencies:
                 self._when_rules_section_dependencies_by_section[
                     section["id"]
                 ] = rules_section_dependencies
 
+            if rule_section_dependencies_for_progress_value_source:
+                self.__when_rules_section_dependencies_by_section_for_progress_value_source[
+                    section["id"]
+                ] = rule_section_dependencies_for_progress_value_source
```

Comment:
There is an issue with this in that, because you are saying the section is dependent even if w it was a block selector then we will be trying to reevaluate the path for every block in the section. So if Section A had 20 blocks, and the 20th block was needed for Section B, C, D, we will currently be trying to reevaluate the path for B, C, D even when the other 19 blocks may be updated, it should only need to re-evaluate if block 20 was updated.

Therefore, you need to capture the dependencies by sections and blocks based on the selector. Then when checking dependency, check if there are any section deps for the current section, if not then check section dependencies for current block.

Reply:
Separated the logic for blocks and sections